src/cache.js

```diff
@@ -0,0 +1,69 @@
+const {readFileSync, createReadStream} = require('fs');
+const tar = require('tar');
+const AWS = require('aws-sdk');
+const {sync: globbySync} = require('globby');
+const tempy = require('tempy');
+
+AWS.config.credentials = new AWS.SharedIniFileCredentials({profile: 'automation-aws'});
+
+const CI_CACHE_BUCKET = `ci-cache`;
+const cacheKey = process.env.NPM_CI_CACHE_KEY || `${process.env.system.repo.owner}__${process.env.system.repo}__${process.env.BRANCH}`;
+
+const s3Client = new AWS.S3();
+
+export async function extractCache() {
+  console.log(`Starting to download and extract cache with key ${cacheKey}...`);
+  s3Client.getObject({
+    Bucket: CI_CACHE_BUCKET,
+    Key: cacheKey
+  }).createReadStream()
+    .pipe(tar.extract({}))
+    .on('error', err => console.log('error while downloading and extracting cache', err))
+    .on('end', () => {
+      console.log(`Finished downloading and extracting cache.`);
+    });
+}
+
+export async function saveCache() {
+  const ciConfig = JSON.parse(readFileSync('.ci_config', 'utf8'));
+
+  if (!ciConfig.cache) {
+    console.log('No cache config in .ci_config. Skipping cache creation.');
+  } else {
+    console.log('Found cache config for the following path globs:');
+    ciConfig.cache.paths.forEach(console.log.bind(console));
+
+    const pathsToCache = globbySync(ciConfig.cache.paths, {
+      onlyFiles: false,
+      expandDirectories: false
+    });
+
+    console.log('Expanded the globs to the following paths to cache:');
+    pathsToCache.forEach(console.log.bind(console));
+
+    const tempFile = tempy.file({extension: 'tar.gz'});
+
+    console.log(`Creating tar.gz of cache paths at ${tempFile}...`);
+    await tar.create({
```

Comment:
Maybe should wrap in `try`/`catch` just in case

Reply:
I was planning on wrapping both commands with `|| exit 0` in the `npmBuild.sh` so I could avoid any kind of error handling in these commands, but now that I checked, I'm not sure I need to call the commands from there. Will update.